Get inodes and disk usage via pure go

ncopa opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

**What type of PR is this?**

/kind bug

**What this PR does / why we need it**:

Use pure go to calculate the disk usage and inodes metrics of a directory instead of executing external tools `du` and `find`.
It also adds a testcase for calculating inodes correct when there are hard links.

**Which issue(s) this PR fixes**:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #95172
Fixes #95186
Fixes #96114

**Special notes for your reviewer**:

I am not sure this needs to be mentioned in release notes.

**Does this PR introduce a user-facing change?**:
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
NONE
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

ncopa committed:
Get inodes and disk usage via pure go

Fix inode usage calculation to use filepath.Walk instead of executing an
external find. Also calculate the disk usage while at it so we also get
rid of the external dependency of `nice` and `du`. (#95172)

This is similar to what cadvisor does since commit
https://github.com/google/cadvisor/commit/046818d64c0af62a4c5037583b467296bb68626d

This solves three problems:
- Counts number of inodes correct when there are hardlinks (#96114)
- Makes kubelet work without GNU findutils (#95186)
- Makes kubelet work without GNU coreutils (#95172)

ncopa committed:
Add test for counting inodes correct with hardlinks

knight42 commented:
/ok-to-test

ncopa commented:
I have set up a windows machine to fix the `pkg/volume/util/fs/fs_windows_test.go`, but there are multiple logical errors so I cannot understand how this has ever worked? It looks like this has only been built, but never actually run?

ncopa commented:
I cleaned up various bugs in windows unit test and tested that the unit test actually runs and passes on windows while at it.

ncopa commented:
/retest

ncopa commented:
/retest

ncopa commented:
The tests passed! :tada:

Would it be an idea to include this in 1.20 release? :pray:

jingxu97 commented on pkg/volume/util/fs/fs_windows_test.go:
```diff
@@ -17,65 +17,71 @@ limitations under the License.
-		t.Fatalf("TestDiskUsage failed: %s", err.Error())
+		t.Fatalf("Lstat failed for %s: %v", file_2, err)
 	}
+
```
based on the GetDisUsage implementation, isn't total should be equal to the returned usage size of dir1 because dir1 is the top level dir?

ncopa replied:
Possibly, but I am not very familiar with how windows disk usage calculation works, and previous test did not test for equal:
```go
    if size.Cmp(used) != 1 {
        t.Fatalf("TestDiskUsage failed: %s", err.Error())
    }
```

I though there might be something I'm not aware of on windows, so I changed it to at least accept if they are equal so the test passes.

But Yeah, I think it should be equal (and it was on my test windows test vm)

jingxu97 replied:
size.Cmp(used) != 1 means    size and used are not equal. So the original test is trying to test if size is not equal to used, it will fail.

ncopa replied:
According the [documentation](https://godoc.org/k8s.io/apimachinery/pkg/api/resource#Quantity.Cmp) `size.Cmp(used) != 1` means that size is not greater than used.

> Cmp returns 0 if the quantity is equal to y, -1 if the quantity is less than y, or 1 if the quantity is greater than y.

jingxu97 replied:
oh, my mistake. So here we should test equal size.Cmp(used) == 0?

ncopa replied:
If the intention was to test the result to be equal, then `size.Cmp(used) == 0`. I can make the refactored test check for equal if you prefer that. I need to rebase it anyway.

ncopa commented:
FWIW, I have tested that the test actually passes on windows. (seems like nobody ever tested that before):
```
natanael copa@DESKTOP-EKHR4MH C:\Users\Natanael Copa\go\src\github.com\kubernetes>go test ./pkg/volume/util/fs       
ok      k8s.io/kubernetes/pkg/volume/util/fs    0.129s

```

ncopa commented:
/retest

ncopa commented:
/retest

ncopa commented:
We currently don't run any unit tests on Linux for disk size calculations. Would it make sense to have a common unit test that works for both Linux and windows?

ncopa commented:
/retest

ncopa commented:
/retest

ncopa commented:
/retest

ncopa commented:
/retest

ncopa commented:
@jingxu97 @saad-ali Is there anything I can do to move this forward?

ncopa commented:
/retest

ncopa commented:
I rebased and simplified the changes so they are less intrusive. I believe the windows tests are still broken, but that should be handled in separate PR.

Any chance to get this in for 1.21?

jingxu97 commented on pkg/volume/testing/testing.go:
```diff
@@ -1229,15 +1229,14 @@ func FindEmptyDirectoryUsageOnTmpfs() (*resource.Quantity, error) {
 	}
 	defer os.RemoveAll(tmpDir)
-	out, err := exec.New().Command("nice", "-n", "19", "du", "-x", "-s", "-B", "1", tmpDir).CombinedOutput()
+	out, err := exec.New().Command("nice", "-n", "19", "du", "-x", "-s", "-k", tmpDir).CombinedOutput()
```
what this change for?

ncopa replied:
Sounds like I need to improve the commit message.

While looking at issue #95172 I grep'ed the code for other calls to `du` and found this place.

It fixes du command to run on Alpine Linux which uses busybox instead of GNU coreutils. (#95172)

The `-B 1` argument is a GNU extension, not defined in POSIX du[1] and does not work with
busybox du used by Alpine Linux. Fix this by using `-k`, which is defined in POSIX[1] and portable and append "Ki" for the ParseQuantity call so du's output is properly parsed.

`resource.ParseQuantity()` will set the format to `BinarySI` when there
is a "Ki" suffix so we can remove the explicit set of the format.

[1]: https://pubs.opengroup.org/onlinepubs/9699919799/utilities/du.html

Maybe I should do this in separate PR? I'm not sure where/when `pkg/volume/testing`  is used, but I will definitively fail on Alpine Linux unless you install coreutils. With this change it will be portable to anything POSIX.

Or maybe we should let it call the go implementation instead to get rid of the external `du` dependency?

jingxu97 replied:
I think go implementation is better. You can either change here or make a separate PR.

ncopa replied:
I had a look at this, but this is only used by the unit tests. I think it makes more sense to call the external `du` to verify that the go implementation in this PR corresponds with what is expected from `du` instead of comparing two identical calls to the same function, so I don't think we should use the go implementation here.

I think my commit is slightly better than what it currently does, but I'm ok to remove it and require GNU coreutils for running the unit tests.

ncopa replied:
actually, lets deal with this in separate PR.

jingxu97 commented:
> I rebased and simplified the changes so they are less intrusive. I believe the windows tests are still broken, but that should be handled in separate PR.
> 
> Any chance to get this in for 1.21?

very sorry I missed this. 1.21 is frozen. But I think we could cherrypick it if it cannot get in. I will review by today.

jingxu97 commented on pkg/volume/util/fs/fs.go:
```diff
@@ -55,63 +59,90 @@ func Info(path string) (int64, int64, int64, int64, int64, int64, error) {
-		return inodes.Value(), nil
-	} else if err != nil {
-		return 0, fmt.Errorf("unable to retrieve inode consumption via quota for %s: %v", path, err)
+		usage.Bytes = consumption.Value()
```
should it be if consumption != nil?

ncopa replied:
That is a good question.

Previously the code looked like this:
```go
	// First check whether the quota system knows about this directory
	// A nil quantity with no error means that the path does not support quotas
	// and we should use other mechanisms.
	data, err := fsquota.GetConsumption(path)
	if data != nil {
		return data, nil
	} else if err != nil {
		return nil, fmt.Errorf("unable to retrieve disk consumption via quota for %s: %v", path, err)
	}
```
But what happens if `fsquota.GetConsuption(path)` returns a data value that is non-nil _and_ an error? With the above code, it would gladly return the `data` even if there was an error returned. 

Normally in Go code, if there is an error returned, the other value(s) returned are invalid and must be discarded, but I'm not sure that is the case with `fsquota.GetConsumption` and `fsquota.GetInodes`? are there any valid case where those return a value _and_ an error, where we should ignore the error?

My change looks like this (which is difficult to see by just looking at the git diff):
```go
    // First check whether the quota system knows about this directory
    // A nil quantity with no error means that the path does not support quotas
    // and we should use other mechanisms.
    consumption, err := fsquota.GetConsumption(path)
    if err != nil {
        return usage, fmt.Errorf("unable to retrieve disk consumption via quota for %s: %v", path, err)
    }
    inodes, err := fsquota.GetInodes(path)
    if err != nil {
        return usage, fmt.Errorf("unable to retrieve inode consumption via quota for %s: %v", path, err)
    }

    if inodes != nil {
        usage.Bytes = consumption.Value()
    }

    if consumption != nil {
        usage.Inodes = inodes.Value()
    }

    if inodes != nil && consumption != nil {
        return usage, nil
    }
```

So, if either of `fsquota.GetConsumption` or `fsquota.GetInodes` returns error, we abort everything with that error, assuming that the return value can not be used.

Once we have tested that the return values are valid (eg no errors returned) we check if they are nil or not, and if both are non-nil we can return without walking the path.

However, if we are supposed to ignore any errors from either `fsquota.GetConsumption` or `fsquota.GetInodes` as long as the return value is non-nil, then my code is wrong and needs change.

jingxu97 replied:
>So, if either of fsquota.GetConsumption or fsquota.GetInodes returns error, we abort everything with that error, assuming that the return value can not be used.

In this case, my thought is we should use the filepath.walk approach to get the size instead of returning error directly. I checked quota function, and I don't think there will be case that data!=nil && err!=nil

```go
    // First check whether the quota system knows about this directory
    // A nil quantity with no error means that the path does not support quotas
    // and we should use other mechanisms.
    consumption, err1 := fsquota.GetConsumption(path)
    inodes, err2 := fsquota.GetInodes(path)

    if consumption != nil && inode !=nil {
             usage.Bytes = consumption.Value()
             usaga.Inodes = inode.Value()
            return usage, nil
    }

     ...

    // the following use filepath.Walk to get usage

   ```

ncopa replied:
I think you are right, looks like `fsquota.GetConsumption` returns error in case `xfs_quota` fails to execute, and we dont want return error in that case. I think we simply want fallback to `filepath.walk` on any error.

jingxu97 commented on pkg/volume/metrics_du.go:
```diff
@@ -64,23 +59,14 @@ func (md *metricsDu) GetMetrics() (*Metrics, error) {
-func (md *metricsDu) runFind(metrics *Metrics) error {
-	inodesUsed, err := fs.Find(md.path)
+// getDiskUsage writes metrics.Used and metric.InodesUsed from fs.DiskUsage
+func (md *metricsDu) getDiskUsage(metrics *Metrics) error {
```
Since no longer use du command, maybe change the file name, function name etc?

ncopa replied:
its still "usage of disk for specified directory metrics". I can rename it but then the change becomes more intrusive. Any suggestions of new name?

jingxu97 replied:
I am ok to leave this to future PR.

ncopa commented on pkg/volume/util/fs/fs.go:
```diff
@@ -55,63 +59,90 @@ func Info(path string) (int64, int64, int64, int64, int64, int64, error) {
-	if err := findCmd.Start(); err != nil {
-		return 0, fmt.Errorf("failed to exec cmd %v - %v; stderr: %v", findCmd.Args, err, stderr.String())
+
+	if consumption != nil {
```
see my response above when  about consumption != nil.

ncopa replied:
> this should inode != nil?

I finally understood what you meant here, and yes, it should, and its fixed now. Thanks!

ncopa commented on pkg/volume/util/fs/fs.go:
```diff
@@ -55,63 +59,90 @@ func Info(path string) (int64, int64, int64, int64, int64, int64, error) {
+			return fmt.Errorf("fileinfo Sys is nil")
+		}
+
+		s, ok := info.Sys().(*syscall.Stat_t)
```
I was in doubt if this is ok myself, but I searched the code and found it to be used various other places. looks like those are linux specific though.

But both [darwin arm64](https://github.com/golang/go/blob/master/src/syscall/ztypes_darwin_arm64.go#L66) and [darwin amd64](https://github.com/golang/go/blob/master/src/syscall/ztypes_darwin_amd64.go#L66) has both `Dev` and `NLink` fields so darwin should be fine.

Looks plan9 would need speical handling indeed. Also does not look like there are any `NLink`s for plan9? Do we need to worry about that now?

ncopa commented:
I want to comment on  using os.Stat, but github thinks im adding a review for some reason and is not letting me respond witha simple comment. ... weird..

We could but then we need the `.Sys().(*syscall.Stat_t)` conversion to get the Stat_t.Dev out from it:

```diff
@@ -91,12 +91,16 @@ func DiskUsage(path string) (UsageInfo, error) {
                return usage, nil
        }
 
-       topLevelStat := &unix.Stat_t{}
-       err = unix.Stat(path, topLevelStat)
+       topLevelInfo, err := os.Stat(path)
        if err != nil {
                return usage, err
        }
 
+       topLevelStat, ok := topLevelInfo.Sys().(*syscall.Stat_t)
+       if !ok {
+               return fmt.Errorf("unsupported fileinfo; could not convert to stat_t")
+       }
+
        // dedupedInode stores inodes that could be duplicates (nlink > 1)
        dedupedInodes := make(map[uint64]struct{})
 
```

I can change it to the above if you prefer, but I found the `unix.Stat` better option, which should work for any POSIX compliant system.

ncopa commented on pkg/volume/util/fs/fs.go:
```diff
@@ -55,63 +59,90 @@ func Info(path string) (int64, int64, int64, int64, int64, int64, error) {
-// Simple io.Writer implementation that counts how many bytes were written.
-type byteCounter struct{ bytesWritten int64 }
+	topLevelStat := &unix.Stat_t{}
+	err = unix.Stat(path, topLevelStat)
```
We could but then we need the `.Sys().(*syscall.Stat_t)` conversion to get the Stat_t.Dev out from it:

```diff
@@ -91,12 +91,16 @@ func DiskUsage(path string) (UsageInfo, error) {
                return usage, nil
        }
 
-       topLevelStat := &unix.Stat_t{}
-       err = unix.Stat(path, topLevelStat)
+       topLevelInfo, err := os.Stat(path)
        if err != nil {
                return usage, err
        }
 
+       topLevelStat, ok := topLevelInfo.Sys().(*syscall.Stat_t)
+       if !ok {
+               return fmt.Errorf("unsupported fileinfo; could not convert to stat_t")
+       }
+
        // dedupedInode stores inodes that could be duplicates (nlink > 1)
        dedupedInodes := make(map[uint64]struct{})
 
```

I can change it to the above if you prefer, but I found the `unix.Stat` better option, which should work for any POSIX compliant system.

ncopa commented:
/retest

ncopa commented:
/retest

ncopa commented:
/retest

ncopa commented:
/retest

ncopa commented:
/retest

ncopa commented:
/test pull-kubernetes-e2e-azure-file-windows

ncopa commented:
/retest

jingxu97 commented:
/approve
/lgtm

jingxu97 commented:
cc @ddebroy

dims commented:
/assign @msau42 @xing-yang

ddebroy commented:
No concerns around the refactoring for Windows. The other changes look good too.

/lgtm

xing-yang commented:
/lgtm
/approve

ncopa commented:
/retest

ncopa commented:
/retest

ncopa commented:
/retest pull-kubernetes-e2e-gce-ubuntu-containerd

ncopa commented:
/test pull-kubernetes-e2e-gce-ubuntu-containerd

ncopa commented:
/retest
